File: packages/oauth-resource-server-ts/src/middleware.ts
import { z } from 'zod';
import { ClaimsSchema, type Claims, hasAllScopes, hasAnyScope } from './claims.js';
import { AuthorizationError, type ErrorCode } from './errors.js';
import { Verifier, bearerToken } from './verifier.js';

/**
 * AuthedRequestSchema is the zod schema for a request carrying an optional
 * resolved claims property (Express/connect-style). The `AuthedRequest` type is
 * inferred from it.
 */
export const AuthedRequestSchema = z.object({
  headers: z.record(z.string(), z.union([z.string(), z.array(z.string())]).optional()),
  method: z.string().optional(),
  url: z.string().optional(),
  body: z.unknown().optional(),
  claims: ClaimsSchema.optional(),
}).passthrough();

/** A request with an optional resolved claims property (Express/connect-style). */
export type AuthedRequest = z.infer<typeof AuthedRequestSchema>;

/**
 * Minimal response surface (compatible with Express). This is a recursive,
 * method-bearing contract describing a framework object rather than validatable
 * data, so the schema is a {@link z.custom} pass-through over the type.
 */
export type ResponseLike = {
  status(code: number): ResponseLike;
  setHeader(name: string, value: string): void;
  json(body: unknown): void;
};
export const ResponseLikeSchema = z.custom<ResponseLike>();

/**
 * NextFn is the Express/connect `next` callback. As a callable (not data), it is
 * represented with {@link z.custom} over the function type.
 */
export type NextFn = (err?: unknown) => void;
export const NextFnSchema = z.custom<NextFn>();

function sendError(res: ResponseLike, err: AuthorizationError): void {
  res.status(err.status);
  if (err.status === 401) res.setHeader('WWW-Authenticate', 'Bearer');
  res.json({ error: err.message, code: err.code });
}

function denial(code: ErrorCode, status: number, message: string, cause?: unknown): AuthorizationError {
  return new AuthorizationError(code, status, message, cause);
}

function headerValue(v: string | string[] | undefined): string | undefined {
  return Array.isArray(v) ? v[0] : v;
}

/**
 * requireAuth is Express/connect-style middleware that verifies the bearer
 * token and attaches `req.claims`. Responds 401 on any failure.
 */
export function requireAuth(verifier: Verifier) {
  return async (req: AuthedRequest, res: ResponseLike, next: NextFn): Promise<void> => {
    const raw = bearerToken(headerValue(req.headers['authorization']));
    if (!raw) {
      sendError(res, denial('token_missing', 401, 'missing bearer token'));
      return;
    }
    try {
      req.claims = await verifier.verify(raw);
      next();
    } catch (err) {
      sendError(res, err instanceof AuthorizationError ? err : denial('token_invalid_signature', 401, 'invalid token signature', err));
    }
  };
}

/**
 * requireScopes enforces that the caller holds every named scope. Mount after
 * requireAuth. Responds 403 on missing scope.
 */
export function requireAllScopes(...scopes: string[]) {
  return (req: AuthedRequest, res: ResponseLike, next: NextFn): void => {
    if (!req.claims) {
      sendError(res, denial('token_missing', 401, 'missing bearer token'));
      return;
    }
    if (!hasAllScopes(req.claims, ...scopes)) {
      sendError(res, denial('scope_missing', 403, 'required scope missing'));
      return;
    }
    next();
  };
}

/** Backward-compatible alias for all-of scope enforcement. */
export const requireScopes = requireAllScopes;

/** Enforces that the caller holds at least one named scope. */
export function requireAnyScope(...scopes: string[]) {
  return (req: AuthedRequest, res: ResponseLike, next: NextFn): void => {
    if (!req.claims) {
      sendError(res, denial('token_missing', 401, 'missing bearer token'));
      return;
    }
    if (scopes.length === 0 || !hasAnyScope(req.claims, ...scopes)) {
      sendError(res, denial('scope_missing', 403, 'required scope missing'));
      return;
    }
    next();
  };
}

export type ConnectionStatusValidator = (claims: Claims) => boolean | Promise<boolean>;
export type ApprovalValidator = (approvalToken: string, claims: Claims, request: AuthedRequest) => boolean | Promise<boolean>;

export type MCPTokenOptions = {
  /** Route-level audience requirement, in addition to verifier configuration. */
  audience?: string;
  /** All-of scope requirements. */
  requiredScopes?: string[];
  /** Any-of scope requirements. */
  anyScopes?: string[];
  /** Optional online connection revocation/status validation. */
  connectionValidator?: ConnectionStatusValidator;
  /** Optional money-moving approval validation/introspection. */
  approvalValidator?: ApprovalValidator;
};

/**
 * Composes RFC 012 bearer, scope, connection, and approval checks. A configured
 * approval validator makes X-Approval-Token mandatory and failures return 428.
 */
export function requireMCPToken(verifier: Verifier, options: MCPTokenOptions = {}) {
  return async (req: AuthedRequest, res: ResponseLike, next: NextFn): Promise<void> => {
    const raw = bearerToken(headerValue(req.headers['authorization']));
    if (!raw) {
      sendError(res, denial('token_missing', 401, 'missing bearer token'));
      return;
    }
    try {
      req.claims = await verifier.verify(raw);
    } catch (err) {
      sendError(res, err instanceof AuthorizationError ? err : denial('token_invalid_signature', 401, 'invalid token signature', err));
      return;
    }
    const claims = req.claims;
    if (options.audience && !claims.audience.includes(options.audience)) {
      sendError(res, denial('audience_mismatch', 401, 'token audience mismatch'));
      return;
    }
    if (options.requiredScopes && !hasAllScopes(claims, ...options.requiredScopes)) {
      sendError(res, denial('scope_missing', 403, 'required scope missing'));
      return;
    }
    if (options.anyScopes && (options.anyScopes.length === 0 || !hasAnyScope(claims, ...options.anyScopes))) {
      sendError(res, denial('scope_missing', 403, 'required scope missing'));
      return;
    }
    if (options.connectionValidator) {
      try {
        if (!(await options.connectionValidator(claims))) {
          sendError(res, denial('connection_revoked', 403, 'connection revoked'));
          return;
        }
      } catch (err) {
        sendError(res, denial('connection_revoked', 403, 'connection revoked', err));
        return;
      }
    }
    if (options.approvalValidator) {
      const token = headerValue(req.headers['x-approval-token'])?.trim();
      if (!token) {
        sendError(res, denial('approval_required', 428, 'approval required'));
        return;
      }
      try {
        if (!(await options.approvalValidator(token, claims, req))) {
          sendError(res, denial('approval_required', 428, 'approval required'));
          return;
        }
      } catch (err) {
        sendError(res, denial('approval_required', 428, 'approval required', err));
        return;
      }
    }
    next();
  };
}

/**
 * verifyAuthorizationHeader is a framework-agnostic helper for Hono, Fastify,
 * etc.: pass the raw Authorization header, get Claims back (throws TokenError).
 *
 *   app.use(async (c, next) => {
 *     const claims = await verifyAuthorizationHeader(verifier, c.req.header('authorization'));
 *     c.set('claims', claims);
 *     await next();
 *   });
 */
export async function verifyAuthorizationHeader(
  verifier: Verifier,
  authorization: string | undefined | null,
): Promise<Claims> {
  const raw = bearerToken(authorization);
  if (!raw) throw new Error('missing bearer token');
  return verifier.verify(raw);
}

File: packages/oauth-resource-server-go/middleware.go
package oauthrs

import (
	"context"
	"encoding/json"
	"net/http"
	"strings"
	"time"
)

type claimsCtxKey struct{}

// WithClaims attaches verified claims to a context.
func WithClaims(ctx context.Context, c *Claims) context.Context {
	return context.WithValue(ctx, claimsCtxKey{}, c)
}

// ClaimsFromContext returns the verified claims attached by Require.
func ClaimsFromContext(ctx context.Context) (*Claims, bool) {
	c, ok := ctx.Value(claimsCtxKey{}).(*Claims)
	return c, ok && c != nil
}

// Require is net/http middleware that extracts the bearer token, verifies it,
// and attaches the resulting Claims to the request context. It responds 401
// (with a {error, code} JSON envelope) on any failure.
func (v *Verifier) Require(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		raw := BearerToken(r)
		if raw == "" {
			writeAuthorizationError(w, authorizationError(CodeTokenMissing, http.StatusUnauthorized, "missing bearer token", nil))
			return
		}
		claims, err := v.Verify(raw)
		if err != nil {
			if authErr, ok := err.(*AuthorizationError); ok {
				writeAuthorizationError(w, authErr)
			} else {
				writeAuthorizationError(w, classifyTokenError(err))
			}
			return
		}
		next.ServeHTTP(w, r.WithContext(WithClaims(r.Context(), claims)))
	})
}

// RequireScopes is middleware that enforces the caller holds every named scope.
// It must be mounted after Require. Responds 403 on missing scope.
func RequireAllScopes(scopes ...string) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			c, ok := ClaimsFromContext(r.Context())
			if !ok {
				writeAuthorizationError(w, authorizationError(CodeTokenMissing, http.StatusUnauthorized, "missing bearer token", nil))
				return
			}
			if !c.HasAllScopes(scopes...) {
				writeAuthorizationError(w, authorizationError(CodeScopeMissing, http.StatusForbidden, "required scope missing", nil))
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

// RequireScopes is a backward-compatible alias for RequireAllScopes.
func RequireScopes(scopes ...string) func(http.Handler) http.Handler {
	return RequireAllScopes(scopes...)
}

// RequireAnyScope enforces that the caller holds at least one named scope.
func RequireAnyScope(scopes ...string) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			c, ok := ClaimsFromContext(r.Context())
			if !ok {
				writeAuthorizationError(w, authorizationError(CodeTokenMissing, http.StatusUnauthorized, "missing bearer token", nil))
				return
			}
			if len(scopes) == 0 || !c.HasAnyScope(scopes...) {
				writeAuthorizationError(w, authorizationError(CodeScopeMissing, http.StatusForbidden, "required scope missing", nil))
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

// ConnectionStatusValidator performs online revocation/status checks.
// It returns true only for an active connection. Errors deny access.
type ConnectionStatusValidator func(context.Context, *Claims) (bool, error)

// ConnectionValidationMode selects the RFC 012 connection-status policy.
// The zero value is ConnectionValidationLive so production use fails closed.
type ConnectionValidationMode string

const (
	// ConnectionValidationLive requires an online status check on every request.
	ConnectionValidationLive ConnectionValidationMode = "live"
	// ConnectionValidationOfflineDevelopment is an explicit development-only
	// escape hatch that accepts only tokens with a tightly bounded issued TTL.
	ConnectionValidationOfflineDevelopment ConnectionValidationMode = "offline-development"

	// MaxOfflineDevelopmentTokenTTL is the largest token TTL accepted by the
	// explicit offline-development connection policy.
	MaxOfflineDevelopmentTokenTTL = 5 * time.Minute
)

// ApprovalValidator validates or introspects a per-invocation approval token.
// It receives the full request so action/resource details can be matched.
type ApprovalValidator func(*http.Request, string, *Claims) (bool, error)

// MCPTokenOptions composes RFC 012 authorization checks. RequiredScopes are
// all-of. AnyScopes are any-of. A configured ApprovalValidator makes
// X-Approval-Token mandatory.
type MCPTokenOptions struct {
	Audience                 string
	RequiredScopes           []string
	AnyScopes                []string
	ConnectionValidationMode ConnectionValidationMode
	ConnectionValidator      ConnectionStatusValidator
	OfflineMaxTokenTTL       time.Duration
	ApprovalValidator        ApprovalValidator
}

// RequireMCPToken verifies the bearer token, applies scope checks, validates
// connection status, and optionally validates X-Approval-Token. Live validation
// is the default and denies every request when ConnectionValidator is absent or
// fails. Offline development must be selected explicitly and requires an issued
// token TTL no larger than both OfflineMaxTokenTTL and five minutes.
func (v *Verifier) RequireMCPToken(opts MCPTokenOptions) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		h := next
		if opts.ApprovalValidator != nil {
			h = requireApproval(opts.ApprovalValidator)(h)
		}
		h = requireConnectionPolicy(opts)(h)
		if len(opts.AnyScopes) > 0 {
			h = RequireAnyScope(opts.AnyScopes...)(h)
		}
		if len(opts.RequiredScopes) > 0 {
			h = RequireAllScopes(opts.RequiredScopes...)(h)
		}
		if opts.Audience != "" {
			h = requireAudience(opts.Audience)(h)
		}
		return v.Require(h)
	}
}

func requireConnectionPolicy(opts MCPTokenOptions) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			c, ok := ClaimsFromContext(r.Context())
			if !ok {
				writeAuthorizationError(w, authorizationError(CodeTokenMissing, http.StatusUnauthorized, "missing bearer token", nil))
				return
			}

			mode := opts.ConnectionValidationMode
			if mode == "" {
				mode = ConnectionValidationLive
			}
			switch mode {
			case ConnectionValidationLive:
				if opts.ConnectionValidator == nil {
					writeAuthorizationError(w, authorizationError(CodeConnectionRevoked, http.StatusForbidden, "active connection validation required", nil))
					return
				}
				active, err := opts.ConnectionValidator(r.Context(), c)
				if err != nil || !active {
					writeAuthorizationError(w, authorizationError(CodeConnectionRevoked, http.StatusForbidden, "connection revoked", err))
					return
				}
			case ConnectionValidationOfflineDevelopment:
				maxTTL := opts.OfflineMaxTokenTTL
				issuedTTL := c.Expiry.Sub(c.IssuedAt)
				if maxTTL <= 0 || maxTTL > MaxOfflineDevelopmentTokenTTL || c.IssuedAt.IsZero() || !c.Expiry.After(c.IssuedAt) || issuedTTL > maxTTL {
					writeAuthorizationError(w, authorizationError(CodeConnectionRevoked, http.StatusForbidden, "offline development token policy denied", nil))
					return
				}
			default:
				writeAuthorizationError(w, authorizationError(CodeConnectionRevoked, http.StatusForbidden, "active connection validation required", nil))
				return
			}

			next.ServeHTTP(w, r)
		})
	}
}

func requireAudience(audience string) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			c, ok := ClaimsFromContext(r.Context())
			if !ok {
				writeAuthorizationError(w, authorizationError(CodeTokenMissing, http.StatusUnauthorized, "missing bearer token", nil))
				return
			}
			matched := false
			for _, actual := range c.Audience {
				if actual == audience {
					matched = true
					break
				}
			}
			if !matched {
				writeAuthorizationError(w, authorizationError(CodeAudienceMismatch, http.StatusUnauthorized, "token audience mismatch", nil))
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

func requireApproval(validate ApprovalValidator) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			c, ok := ClaimsFromContext(r.Context())
			if !ok {
				writeAuthorizationError(w, authorizationError(CodeTokenMissing, http.StatusUnauthorized, "missing bearer token", nil))
				return
			}
			token := strings.TrimSpace(r.Header.Get("X-Approval-Token"))
			if token == "" {
				writeAuthorizationError(w, authorizationError(CodeApprovalRequired, http.StatusPreconditionRequired, "approval required", nil))
				return
			}
			valid, err := validate(r, token, c)
			if err != nil || !valid {
				writeAuthorizationError(w, authorizationError(CodeApprovalRequired, http.StatusPreconditionRequired, "approval required", err))
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

// BearerToken extracts the token from an "Authorization: Bearer <token>"
// header, returning "" if absent or malformed.
func BearerToken(r *http.Request) string {
	h := r.Header.Get("Authorization")
	if h == "" {
		return ""
	}
	const prefix = "Bearer "
	if len(h) > len(prefix) && strings.EqualFold(h[:len(prefix)], prefix) {
		return strings.TrimSpace(h[len(prefix):])
	}
	return ""
}

func writeAuthorizationError(w http.ResponseWriter, err *AuthorizationError) {
	w.Header().Set("Content-Type", "application/json")
	if err.Status == http.StatusUnauthorized {
		w.Header().Set("WWW-Authenticate", "Bearer")
	}
	w.WriteHeader(err.Status)
	_ = json.NewEncoder(w).Encode(map[string]string{"error": err.Message, "code": string(err.Code)})
}
